File: server/src/domain/revision.ts
/**
 * Conflict policy for mutable entities (Product, Customer, and any future
 * Shop settings). Append-only ledger entities (Sale, StockMovement,
 * BakiEntry) never go through this — they're never edited in place.
 * See DECISIONS.md D017.
 */

/** Language-neutral status code — see DECISIONS.md D011. Never localize this string itself. */
export const REVISION_CONFLICT = 'REVISION_CONFLICT' as const

export interface Revisioned<T> {
  readonly data: T
  readonly revision: number
  readonly updatedAt: string
  readonly deletedAt: string | null
}

export type RevisionCheckResult =
  { readonly ok: true } | { readonly ok: false; readonly code: typeof REVISION_CONFLICT }

/**
 * A client write must include the revision it edited from (baseRevision).
 * If it no longer matches the server's current revision, the write is
 * rejected — never silently overwritten.
 */
export function checkRevision(currentRevision: number, baseRevision: number): RevisionCheckResult {
  if (currentRevision !== baseRevision) {
    return { ok: false, code: REVISION_CONFLICT }
  }
  return { ok: true }
}

export function createRevisioned<T>(data: T, createdAt: Date): Revisioned<T> {
  return { data, revision: 1, updatedAt: createdAt.toISOString(), deletedAt: null }
}

/**
 * Applies an update only if baseRevision still matches. On success, bumps
 * the revision and updatedAt. On conflict, returns the conflict result and
 * leaves the entity untouched.
 */
export function applyRevisionedUpdate<T>(
  current: Revisioned<T>,
  baseRevision: number,
  newData: T,
  updatedAt: Date,
): { readonly result: RevisionCheckResult; readonly next: Revisioned<T> } {
  const result = checkRevision(current.revision, baseRevision)
  if (!result.ok) {
    return { result, next: current }
  }
  return {
    result,
    next: {
      data: newData,
      revision: current.revision + 1,
      updatedAt: updatedAt.toISOString(),
      deletedAt: current.deletedAt,
    },
  }
}

/**
 * Deletion is itself a revisioned, conflict-checked write — it sets a
 * tombstone rather than removing the row, so deletion can be synced like
 * any other change.
 */
export function applyRevisionedDelete<T>(
  current: Revisioned<T>,
  baseRevision: number,
  deletedAt: Date,
): { readonly result: RevisionCheckResult; readonly next: Revisioned<T> } {
  const result = checkRevision(current.revision, baseRevision)
  if (!result.ok) {
    return { result, next: current }
  }
  return {
    result,
    next: {
      data: current.data,
      revision: current.revision + 1,
      updatedAt: deletedAt.toISOString(),
      deletedAt: deletedAt.toISOString(),
    },
  }
}

File: android/app/src/main/java/com/hisab/app/domain/Revision.kt
package com.hisab.app.domain

import java.time.Instant

// Conflict policy for mutable entities (Product, Customer, and any future
// Shop settings). Append-only ledger entities (Sale, StockMovement,
// BakiEntry) never go through this — they're never edited in place.
// See DECISIONS.md D017.

/** Language-neutral status code — see DECISIONS.md D011. Never localize this string itself. */
const val REVISION_CONFLICT = "REVISION_CONFLICT"

sealed interface RevisionCheckResult {
    data object Ok : RevisionCheckResult

    data object Conflict : RevisionCheckResult {
        const val CODE = REVISION_CONFLICT
    }
}

data class Revisioned<T>(
    val data: T,
    val revision: Int,
    val updatedAt: Instant,
    val deletedAt: Instant?,
)

/**
 * A client write must include the revision it edited from (baseRevision).
 * If it no longer matches the server's current revision, the write is
 * rejected — never silently overwritten.
 */
fun checkRevision(
    currentRevision: Int,
    baseRevision: Int,
): RevisionCheckResult =
    if (currentRevision != baseRevision) {
        RevisionCheckResult.Conflict
    } else {
        RevisionCheckResult.Ok
    }

fun <T> createRevisioned(
    data: T,
    createdAt: Instant,
): Revisioned<T> = Revisioned(data = data, revision = 1, updatedAt = createdAt, deletedAt = null)

data class RevisionedWrite<T>(
    val result: RevisionCheckResult,
    val next: Revisioned<T>,
)

/**
 * Applies an update only if baseRevision still matches. On success, bumps
 * the revision and updatedAt. On conflict, returns the conflict result and
 * leaves the entity untouched.
 */
fun <T> applyRevisionedUpdate(
    current: Revisioned<T>,
    baseRevision: Int,
    newData: T,
    updatedAt: Instant,
): RevisionedWrite<T> {
    val result = checkRevision(current.revision, baseRevision)
    if (result !is RevisionCheckResult.Ok) {
        return RevisionedWrite(result, current)
    }
    return RevisionedWrite(
        result,
        current.copy(data = newData, revision = current.revision + 1, updatedAt = updatedAt),
    )
}

/**
 * Deletion is itself a revisioned, conflict-checked write — it sets a
 * tombstone rather than removing the row, so deletion can be synced like
 * any other change.
 */
fun <T> applyRevisionedDelete(
    current: Revisioned<T>,
    baseRevision: Int,
    deletedAt: Instant,
): RevisionedWrite<T> {
    val result = checkRevision(current.revision, baseRevision)
    if (result !is RevisionCheckResult.Ok) {
        return RevisionedWrite(result, current)
    }
    return RevisionedWrite(
        result,
        current.copy(revision = current.revision + 1, updatedAt = deletedAt, deletedAt = deletedAt),
    )
}
